File: clients/android/app/src/main/java/com/atlasbot/client/auth/AuthSession.kt
package com.atlasbot.client.auth

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicReference

/**
 * M1 / WM1 Android login: Chrome Custom Tabs.
 * OIDC uses PKCE S256; WeCom uses state + one-time code (no PKCE) → Hub exchange.
 * Primary path is **system browser / Custom Tabs** — not an in-app WebView.
 *
 * Redirect: [MOBILE_REDIRECT_URI] (`atlasbot://auth/callback`).
 * Pending is typed separately ([OidcPending] vs [WeComPending]) so providers cannot mix.
 */
class AuthSession(
    private val tokenStore: TokenStore,
    private val ioExecutor: Executor = Executors.newSingleThreadExecutor(),
) {
    /** OIDC pending — holds PKCE verifier; must not be used for WeCom callback. */
    data class OidcPending(
        val cfg: OidcClientConfig,
        val redirectUri: String,
        val state: String,
        val pkce: PkcePair,
    )

    /** WeCom pending — no PKCE; Hub exchange uses code + optional state. */
    data class WeComPending(
        val cfg: WeComAuth.Config,
        val state: String,
    )

    sealed class Pending {
        data class Oidc(val value: OidcPending) : Pending()
        data class WeCom(val value: WeComPending) : Pending()
    }

    private val pending = AtomicReference<Pending?>(null)
    @Volatile private var callbackListener: ((Result<String>) -> Unit)? = null

    fun hasTicket(): Boolean = !tokenStore.loadBearer().isNullOrBlank()
    fun currentBearer(): String? = tokenStore.loadBearer()
    fun currentProvider(): String? = tokenStore.loadProvider()

    fun saveManualBearer(token: String, provider: String? = null) {
        tokenStore.saveBearer(token.trim(), provider)
    }

    fun logout() {
        pending.set(null)
        tokenStore.clear()
    }

    /**
     * Start OIDC authorize in a Custom Tab. Completion arrives via [handleCallbackUri]
     * after the IdP redirects to [MOBILE_REDIRECT_URI].
     */
    fun startLogin(
        context: Context,
        cfg: OidcClientConfig,
        redirectUri: String = MOBILE_REDIRECT_URI,
        onComplete: (Result<String>) -> Unit,
    ) {
        callbackListener = onComplete
        val pkce = generatePkce()
        val state = randomState()
        pending.set(Pending.Oidc(OidcPending(cfg, redirectUri, state, pkce)))
        val url = buildAuthorizeUrl(cfg, redirectUri, state, pkce)
        launchCustomTab(context, url)
    }

    /**
     * Start WeCom authorize in a Custom Tab (no PKCE).
     * Callback → Hub `POST /auth/wecom/exchange` → TokenStore (provider=wecom).
     */
    fun startWeComLogin(
        context: Context,
        cfg: WeComAuth.Config,
        onComplete: (Result<String>) -> Unit,
    ) {
        callbackListener = onComplete
        val state = randomState()
        pending.set(Pending.WeCom(WeComPending(cfg, state)))
        val url = WeComAuth.buildAuthorizeUrl(cfg, state)
        launchCustomTab(context, url)
    }

    /**
     * Thin dispatcher: `provider=wecom` → [startWeComLogin]; else OIDC [startLogin].
     * Avoids duplicating deep-link state machine in UI.
     */
    fun startLogin(
        context: Context,
        provider: String,
        oidc: OidcClientConfig? = null,
        wecom: WeComAuth.Config? = null,
        redirectUri: String = MOBILE_REDIRECT_URI,
        onComplete: (Result<String>) -> Unit,
    ) {
        when (ticketProviderFromEnv(provider)) {
            WeComAuth.PROVIDER -> {
                val cfg = wecom ?: throw IllegalArgumentException("wecom config required")
                startWeComLogin(context, cfg, onComplete)
            }
            else -> {
                val cfg = oidc ?: throw IllegalArgumentException("oidc config required")
                startLogin(context, cfg, redirectUri, onComplete)
            }
        }
    }

    /** Install WeCom pending without launching a browser (unit tests). */
    fun prepareWeComLogin(cfg: WeComAuth.Config, state: String = randomState()): String {
        pending.set(Pending.WeCom(WeComPending(cfg, state)))
        return state
    }

    /** Install OIDC pending without launching a browser (unit tests). */
    fun prepareOidcLogin(
        cfg: OidcClientConfig,
        redirectUri: String = MOBILE_REDIRECT_URI,
        state: String = randomState(),
        pkce: PkcePair = generatePkce(),
    ): String {
        pending.set(Pending.Oidc(OidcPending(cfg, redirectUri, state, pkce)))
        return state
    }

    private fun launchCustomTab(context: Context, url: String) {
        val intent = CustomTabsIntent.Builder().build()
        intent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.launchUrl(context, Uri.parse(url))
    }

    /** Called from MainActivity when `atlasbot://auth/callback` arrives. */
    fun handleCallbackUri(uri: Uri, onMain: (Result<String>) -> Unit) {
        handleCallbackUrl(uri.toString(), onMain)
    }

    /** String entry for JVM unit tests (no android.net.Uri). */
    fun handleCallbackUrl(callbackUrl: String, onMain: (Result<String>) -> Unit) {
        val p = pending.getAndSet(null)
        if (p == null) {
            onMain(Result.failure(IllegalStateException("no pending login")))
            return
        }
        val parsed = parseCallbackUrl(callbackUrl)
        if (!parsed.error.isNullOrEmpty()) {
            fail(Result.failure(IllegalStateException(parsed.error)), onMain)
            return
        }
        if (parsed.code.isNullOrEmpty()) {
            fail(Result.failure(IllegalStateException("missing code")), onMain)
            return
        }
        when (p) {
            is Pending.Oidc -> handleOidcCallback(p.value, parsed, onMain)
            is Pending.WeCom -> handleWeComCallback(p.value, parsed, onMain)
        }
    }

    private fun handleOidcCallback(
        p: OidcPending,
        parsed: CallbackParse,
        onMain: (Result<String>) -> Unit,
    ) {
        if (parsed.state != null && parsed.state != p.state) {
            fail(Result.failure(IllegalStateException("state mismatch")), onMain)
            return
        }
        val code = parsed.code!!
        ioExecutor.execute {
            try {
                val tr = exchangeCode(p.cfg, p.redirectUri, code, p.pkce.verifier)
                val bearer = pickHubBearer(tr)
                    ?: throw IllegalStateException("no id_token/access_token")
                tokenStore.saveBearer(bearer, provider = "oidc")
                succeed(Result.success(bearer), onMain)
            } catch (t: Throwable) {
                fail(Result.failure(t), onMain)
            }
        }
    }

    private fun handleWeComCallback(
        p: WeComPending,
        parsed: CallbackParse,
        onMain: (Result<String>) -> Unit,
    ) {
        if (parsed.state != null && parsed.state != p.state) {
            fail(Result.failure(IllegalStateException("state mismatch")), onMain)
            return
        }
        val code = parsed.code!!
        ioExecutor.execute {
            try {
                val tr = WeComAuth.exchangeCode(
                    hubHttpBase = p.cfg.hubHttpBase,
                    code = code,
                    state = p.state,
                )
                tokenStore.saveBearer(tr.accessToken, provider = WeComAuth.PROVIDER)
                succeed(Result.success(tr.accessToken), onMain)
            } catch (t: Throwable) {
                fail(Result.failure(t), onMain)
            }
        }
    }

    private fun succeed(ok: Result<String>, onMain: (Result<String>) -> Unit) {
        callbackListener?.invoke(ok)
        callbackListener = null
        onMain(ok)
    }

    private fun fail(err: Result<String>, onMain: (Result<String>) -> Unit) {
        callbackListener?.invoke(err)
        callbackListener = null
        onMain(err)
    }

    companion object {
        /** Process-wide session used by MainActivity deep-link + Compose UI. */
        @Volatile var shared: AuthSession? = null

        fun obtain(context: Context): AuthSession {
            shared?.let { return it }
            return synchronized(this) {
                shared ?: AuthSession(EncryptedPrefsTokenStore(context.applicationContext)).also {
                    shared = it
                }
            }
        }
    }
}

File: clients/ios/AtlasBot/AuthSession.swift
import Foundation
import AuthenticationServices

#if canImport(UIKit)
import UIKit
#endif

/**
 M1 / WM1 iOS login: `ASWebAuthenticationSession`.
 OIDC uses PKCE S256; WeCom uses state + one-time code (no PKCE) → Hub exchange.
 Primary path is the **system browser session** — not an in-app WebView.

 Redirect: `atlasbot://auth/callback` (`callbackURLScheme` = `atlasbot`).
 Pending is typed separately (Oidc vs WeCom) so providers cannot mix.
 */
@MainActor
public final class AuthSession: NSObject {
    public static let shared = AuthSession()

    private enum Pending {
        case oidc(cfg: OidcClientConfig, redirect: String, state: String, pkce: PkcePair)
        case wecom(cfg: WeComAuth.Config, state: String)
    }

    private let tokenStore: TokenStore
    private var webAuthSession: ASWebAuthenticationSession?
    private var pending: Pending?
    private var urlSession: URLSession

    public init(tokenStore: TokenStore = KeychainTokenStore(), urlSession: URLSession = .shared) {
        self.tokenStore = tokenStore
        self.urlSession = urlSession
    }

    public func hasTicket() -> Bool {
        !(tokenStore.loadBearer() ?? "").isEmpty
    }

    public func currentBearer() -> String? { tokenStore.loadBearer() }

    public func currentProvider() -> String? { tokenStore.loadProvider() }

    public func saveManualBearer(_ token: String, provider: String? = nil) {
        tokenStore.saveBearer(token.trimmingCharacters(in: .whitespacesAndNewlines), provider: provider)
    }

    public func logout() {
        pending = nil
        webAuthSession?.cancel()
        webAuthSession = nil
        tokenStore.clear()
    }

    public func startLogin(
        cfg: OidcClientConfig,
        redirectUri: String = MOBILE_REDIRECT_URI,
        presentationAnchor: ASPresentationAnchor? = nil,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        let pkce = generatePkce()
        let state = randomState()
        pending = .oidc(cfg: cfg, redirect: redirectUri, state: state, pkce: pkce)
        let url = buildAuthorizeUrl(cfg: cfg, redirectUri: redirectUri, state: state, pkce: pkce)
        launchWebAuth(url: url, presentationAnchor: presentationAnchor, completion: completion)
    }

    /// Start WeCom authorize (no PKCE). Callback → Hub exchange → TokenStore (provider=wecom).
    public func startWeComLogin(
        cfg: WeComAuth.Config,
        presentationAnchor: ASPresentationAnchor? = nil,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        let state = randomState()
        pending = .wecom(cfg: cfg, state: state)
        guard let url = WeComAuth.buildAuthorizeURL(cfg: cfg, state: state) else {
            completion(.failure(NSError(domain: "wecom", code: 10, userInfo: [NSLocalizedDescriptionKey: "bad authorize url"])))
            return
        }
        launchWebAuth(url: url, presentationAnchor: presentationAnchor, completion: completion)
    }

    /// Thin dispatcher: `provider=wecom` → startWeComLogin; else OIDC startLogin.
    public func startLogin(
        provider: String,
        oidc: OidcClientConfig? = nil,
        wecom: WeComAuth.Config? = nil,
        redirectUri: String = MOBILE_REDIRECT_URI,
        presentationAnchor: ASPresentationAnchor? = nil,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        if ticketProviderFromEnv(provider) == WeComAuth.provider {
            guard let wecom else {
                completion(.failure(NSError(domain: "wecom", code: 11, userInfo: [NSLocalizedDescriptionKey: "wecom config required"])))
                return
            }
            startWeComLogin(cfg: wecom, presentationAnchor: presentationAnchor, completion: completion)
        } else {
            guard let oidc else {
                completion(.failure(NSError(domain: "oidc", code: 11, userInfo: [NSLocalizedDescriptionKey: "oidc config required"])))
                return
            }
            startLogin(cfg: oidc, redirectUri: redirectUri, presentationAnchor: presentationAnchor, completion: completion)
        }
    }

    /// Install WeCom pending without launching a browser (unit tests).
    public func prepareWeComLogin(cfg: WeComAuth.Config, state: String = randomState()) -> String {
        pending = .wecom(cfg: cfg, state: state)
        return state
    }

    /// Install OIDC pending without launching a browser (unit tests).
    public func prepareOidcLogin(
        cfg: OidcClientConfig,
        redirectUri: String = MOBILE_REDIRECT_URI,
        state: String = randomState(),
        pkce: PkcePair = generatePkce()
    ) -> String {
        pending = .oidc(cfg: cfg, redirect: redirectUri, state: state, pkce: pkce)
        return state
    }

    private func launchWebAuth(
        url: URL,
        presentationAnchor: ASPresentationAnchor?,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        let session = ASWebAuthenticationSession(
            url: url,
            callbackURLScheme: MOBILE_CALLBACK_SCHEME
        ) { [weak self] callbackURL, error in
            Task { @MainActor in
                guard let self else { return }
                if let error {
                    completion(.failure(error))
                    return
                }
                guard let callbackURL else {
                    completion(.failure(NSError(domain: "oidc", code: 2, userInfo: [NSLocalizedDescriptionKey: "missing callback"])))
                    return
                }
                self.finishLogin(callbackURL: callbackURL, completion: completion)
            }
        }
        session.prefersEphemeralWebBrowserSession = false
        #if canImport(UIKit)
        if let anchor = presentationAnchor {
            session.presentationContextProvider = AnchorProvider(anchor: anchor)
        } else if let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
                  let window = scene.windows.first(where: \.isKeyWindow) ?? scene.windows.first {
            session.presentationContextProvider = AnchorProvider(anchor: window)
        }
        #endif
        webAuthSession = session
        if !session.start() {
            completion(.failure(NSError(domain: "oidc", code: 3, userInfo: [NSLocalizedDescriptionKey: "failed to start ASWebAuthenticationSession"])))
        }
    }

    /// Optional deep-link entry (URL Types) if session callback is delivered via `onOpenURL`.
    public func handleCallbackURL(_ url: URL, completion: @escaping (Result<String, Error>) -> Void) {
        guard url.scheme == MOBILE_CALLBACK_SCHEME else {
            completion(.failure(NSError(domain: "oidc", code: 4, userInfo: [NSLocalizedDescriptionKey: "unexpected scheme"])))
            return
        }
        finishLogin(callbackURL: url, completion: completion)
    }

    private func finishLogin(callbackURL: URL, completion: @escaping (Result<String, Error>) -> Void) {
        guard let pending else {
            completion(.failure(NSError(domain: "oidc", code: 5, userInfo: [NSLocalizedDescriptionKey: "no pending login"])))
            return
        }
        let parsed = parseCallbackUrl(callbackURL.absoluteString)
        if let err = parsed.error {
            completion(.failure(NSError(domain: "oidc", code: 6, userInfo: [NSLocalizedDescriptionKey: err])))
            return
        }
        guard let code = parsed.code, !code.isEmpty else {
            completion(.failure(NSError(domain: "oidc", code: 7, userInfo: [NSLocalizedDescriptionKey: "missing code"])))
            return
        }
        switch pending {
        case .oidc(let cfg, let redirect, let expectedState, let pkce):
            if let st = parsed.state, st != expectedState {
                completion(.failure(NSError(domain: "oidc", code: 8, userInfo: [NSLocalizedDescriptionKey: "state mismatch"])))
                return
            }
            exchangeCode(cfg: cfg, redirectUri: redirect, code: code, codeVerifier: pkce.verifier) { [weak self] result in
                Task { @MainActor in
                    switch result {
                    case .failure(let e):
                        completion(.failure(e))
                    case .success(let tr):
                        guard let bearer = pickHubBearer(tr) else {
                            completion(.failure(NSError(domain: "oidc", code: 9, userInfo: [NSLocalizedDescriptionKey: "no id_token/access_token"])))
                            return
                        }
                        self?.tokenStore.saveBearer(bearer, provider: "oidc")
                        self?.pending = nil
                        completion(.success(bearer))
                    }
                }
            }
        case .wecom(let cfg, let expectedState):
            if let st = parsed.state, st != expectedState {
                completion(.failure(NSError(domain: "wecom", code: 8, userInfo: [NSLocalizedDescriptionKey: "state mismatch"])))
                return
            }
            WeComAuth.exchangeCode(
                hubHttpBase: cfg.hubHttpBase,
                code: code,
                state: expectedState,
                session: urlSession
            ) { [weak self] result in
                Task { @MainActor in
                    switch result {
                    case .failure(let e):
                        completion(.failure(e))
                    case .success(let tr):
                        self?.tokenStore.saveBearer(tr.accessToken, provider: WeComAuth.provider)
                        self?.pending = nil
                        completion(.success(tr.accessToken))
                    }
                }
            }
        }
    }
}

#if canImport(UIKit)
private final class AnchorProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
    let anchor: ASPresentationAnchor
    init(anchor: ASPresentationAnchor) { self.anchor = anchor }
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor { anchor }
}
#endif
